### PR TITLE
More types & some refactoring

### PR DESCRIPTION
**Pull request by ydylla**
This pull request adds support for more types and also utilizes the code generation capabilities for them.

The following types where added:
* [uuid.UUID](https://docs.python.org/3/library/uuid.html)
* PosixPath, WindowsPath, PurePath, PurePosixPath, PureWindowsPath from [pathlib](https://docs.python.org/3/library/pathlib.html)
* IPv4Address, IPv6Address, IPv4Network, IPv6Network, IPv4Interface, IPv6Interface from [ipaddress](https://docs.python.org/3/library/ipaddress.html)
* date & datetime from [datetime](https://docs.python.org/3/library/datetime.html)

I needed these types because I wanted to use pyserde in combination with [asyncpg](https://github.com/MagicStack/asyncpg).
asyncpg returns dict like objects which can be converted to dataclasses by pyserde's `from_dict` function.

Since asyncpg also supports the above mentioned types, the dicts already contain instances of them.
That is the reason why I also added a new argument `reuse_instances` for `from_dict` & `from_tuple`.
When this is set to `True` these functions check if the field already contains an instances of the target type and reuse it when possible.

This is faster for Path & IPAddress then calling the constructor again.
For UUID it is also the only way to handle existing instances because `uuid.UUID()` does not accept them as an argument.
It is also possible to change the default value of `reuse_instances` via the `serialize` & `deserialize` decorators.

To not cause slowdowns when serializing or deserializing to json/msgpack/toml/yaml `reuse_instances` is always set to `False` there, because we will never see existing instances there.

That is the part where this pull request drifted into refactoring.
Since I had to add the `reuse_instances` argument at all these places, I also used the opportunity to remove some unused arguments (`named` & `strict`) and renamed `asdict` to `to_dict` and `astuple` to `to_tuple`.
These are breaking changes, but I felt it was worth the gained similarity between serialization and deserialization code.
In my option the public exposed functions are now also named more uniformly.

For msgpack external types I also changed the behaviour slightly. It is not required anymore that the dataclasses have a special meaning `_type` attribute.
Instead `to_msgpack` & `from_msgpack` search the `ext_dict` for the correct type or type code and also throw exceptions if they can not find them.

I am sorry that this became so intertwined, I understand if you don't want to merge the breaking changes.
I tried to make the commits cherry-pickable so maybe you only want to use some commits.

During development, I also noticed that Unions do not work properly for these types.
I will make a separate pull request to fix that.

Finally, I have a question: What is the purpose of setting `SE_NAME` and why is it used in `is_serializable`?
Could it be removed? `is_serializable` could use `TO_ITER` and `TO_DICT` for the check like `is_deserializable`.

### CODE REVIEW

**Comment by ydylla** [comment]
I totally forgot that this also adds support for date & datetime from [datetime](https://docs.python.org/3/library/datetime.html)

**Comment by yukinarit** [comment]
Hi @ydylla, 

Thank you for the huge contribution! Let me check little by little 👍

**Comment by yukinarit** [comment]
@ydylla  

> Finally, I have a question: What is the purpose of setting SE_NAME and why is it used in is_serializable?
Could it be removed? is_serializable could use TO_ITER and TO_DICT for the check like is_deserializable.

`SE_NAME`  was used because it was just added before `TO_ITER` and `TO_DICT` as far as remember. You can remove it and use `TO_ITER` and `TO_DICT` instead 🙂

**Review by yukinarit — COMMENTED** [review]
@ydylla 
I have reviewed the first half of commits. 
Please check the review commits, thank you!

**Comment by ydylla** [comment]
> `SE_NAME` was used because it was just added before `TO_ITER` and `TO_DICT` as far as remember. You can remove it and use `TO_ITER` and `TO_DICT` instead 🙂
Okay I will change it. I also guessed this is the reason 😄

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,5 @@ coverage.xml
 html/
 pip-wheel-metadata
 .DS_Store
+venv/
+.idea/
```
[the remark this answers is not in the document]
Comment (ydylla) [reply .gitignore:17]: Sure

File: serde/__init__.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
     'is_deserializable',
     'from_dict',
     'from_tuple',
-    'asdict',
-    'astuple',
+    'to_dict',
+    'to_tuple',
```
[the remark this answers is not in the document]
Comment (ydylla) [reply serde/__init__.py:20]: I actually never used the dataclasses `asdict` & `astuple` functions.
I just found it strange that `asdict` and `to_dict` from master basically did the same thing and wondered why `to_dict` was not exported under `serde`.

My main reason for renaming was to simplify the code and make the serializing & deserializing code more similar (have more symmetry).
For me it felt more natural that the inverse of `from_dict` is `to_dict` especially because the functions of all the other formats follow the same naming schema with `to_` & `from_`.

File: serde/msgpack.py [hunk 3]
```diff
@@ -1,52 +1,57 @@
...
-            maybe_ext = msgpack.ExtType(int(obj._type), obj_bytes)
+    def serialize(cls, obj, use_bin_type: bool = True, ext_type_code: int = None, **opts) -> bytes:
+        if ext_type_code is not None:
+            obj_bytes = msgpack.packb(obj, use_bin_type=use_bin_type, **opts)
+            obj_or_ext = msgpack.ExtType(ext_type_code, obj_bytes)
         else:
-            maybe_ext = asf(obj)
-        return msgpack.packb(maybe_ext, use_bin_type=use_bin_type, **opts)
+            obj_or_ext = obj
+        return msgpack.packb(obj_or_ext, use_bin_type=use_bin_type, **opts)
```
[the remark this answers is not in the document]
Comment (adsharma) [reply serde/msgpack.py:21]: Tested with a raft implementation that depends on ext_dict. Everything looks ok!

File: serde/msgpack.py [hunk 4]
```diff
@@ -1,52 +1,57 @@
...
-def to_msgpack(obj: Any, se=MsgPackSerializer, **opts) -> bytes:
+def to_msgpack(obj: Any, se: Serializer = MsgPackSerializer, named: bool = True, ext_dict: Dict[int, Type] = None, **opts) -> bytes:
     """
     If `ext_dict` option is specified, `obj` is encoded as a `msgpack.ExtType`
     """
-    return se.serialize(obj, **opts)
+    ext_type_code = None
+    if ext_dict is not None:
+        obj_type = type(obj)
+        ext_type_code = next((code for code, ext_type in ext_dict.items() if obj_type is ext_type), None)
```
[the remark this answers is not in the document]
Comment (ydylla) [reply serde/msgpack.py:37]: I know these kind of one liners are bad practice.
But I wanted both functions to accept the same `ext_dict: Dict[int, Type]` (like in the test).
And for `to_msgpack` we have to get the key (integer type code) of the dict by searching with its value (the type).
A simple `ext_dict.get(` only works for the other direction. See `from_msgpack`.

File: serde/msgpack.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yukinarit) [reply serde/msgpack.py:21]: Thank you! 👍

File: serde/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yukinarit) [reply serde/__init__.py:20]: Thanks for the explanation! I am curious of other people's opinion 🙂 

@adsharma @jfuechsl @andreymal  @alexmisk @pranavvp10 
There is a proposal to renaming `asdict`/`astuple` to `to_dict`/`to_tuple`
What do you guys think? 🤔

**Comment by yukinarit** [comment]
@ydylla  
Also, please update [supported types](https://github.com/yukinarit/pyserde#supported-types) section in README.md

**Comment by adsharma** [comment]
I'm neutral. No big deal either way.

File: serde/de.py [hunk 5]
```diff
@@ -379,7 +379,7 @@ def opt(self, arg: DeField) -> str:
                 exists = f'{arg.data} is not None'
             else:
                 exists = f'{arg.datavar}.get("{arg.name}") is not None'
-            return f'{self.render(value)} if {exists} else None'
+            return f'({self.render(value)}) if {exists} else None'
```
[the remark this answers is not in the document]
Comment (ydylla) [reply serde/de.py:399]: Yes, to ensure the evaluation order of the potential nested if else statements. Otherwise the `None` check does not work correctly when the output of `render()` is for a complex type. For example for `Path`.

Without the parenthesis the generated code would look like this:
````python
(data["t"] if isinstance(data["t"], Path) else Path(data["t"]))
if reuse_instances
else Path(data["t"])
if data.get("t") is not None
else None
````
When `reuse_instances` is `True` the None check is skipped and the first line is executed directly, resulting in a TypeError when `data["t"]` equals `None`.

With the parenthesis the code looks like this:
````python
(
    (data["t"] if isinstance(data["t"], Path) else Path(data["t"]))
    if reuse_instances
    else Path(data["t"])
)
if data.get("t") is not None
else None
````
Here the `None` check is executed first and would correctly return `None` if `data["t"]` equals `None`.

**Comment by ydylla** [comment]
> Also, please update [supported types](https://github.com/yukinarit/pyserde#supported-types) section in README.md

Done, I found it to verbose to add all types of `ipaddress` & `pathlib` separately. But feel free to change it if you want a longer list 😃

**Comment by jfuechsl** [comment]
I'd say it's a question of preference. `as*` has a more declarational conotation (which I like) and `to_*` is more imperative (which would be technically "more correct").
If you decide to go with `to_*`, I would ask to keep the original names as aliases for backwards compatibility.

**Comment by yukinarit** [comment]
@adsharma @jfuechsl  
Thanks for your input!

> I'd say it's a question of preference. as* has a more declarational conotation (which I like) and to_* is more imperative (which would be technically "more correct").

Yeah, from_dict/to_dict naming sounds more consistent with other formats but renaming at this time is confusing for the exiting users (although pyserde user base is still small). 

@ydylla 
Thank you for the good suggestion, but I would like not to change.  Could you please remove [the commit](https://github.com/yukinarit/pyserde/pull/80/commits/ed675472341d6f123989cbaacc73071b44f59ea6)?

File: serde/msgpack.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adsharma) [reply serde/msgpack.py:37]: On second thoughts I think it would be more efficient if the `from_msgpack` accepted a reversed dict. That way, the app would do the one time reversal and lookups would be efficient in both directions.

**Comment by ydylla** [comment]
I readded `asdict` & `astuple` in https://github.com/yukinarit/pyserde/pull/80/commits/ae28a33c92d83a24ac718870b8d9d34bc4fa48be with their old signatures and behavior.
I did not revert the commit because I wanted to keep the `to_obj` function, it simplifies the code.
Another reason is in principle you seam to agree with me, so I would suggest to keep the to_* functions and also document them as the new default one.

File: serde/msgpack.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ydylla) [reply serde/msgpack.py:37]: I think you meant `to_msgpack` because `from_msgpack` already uses a simple lookup.

Yes two dicts would be more efficient/faster. Another way could be to add the `ext_type_code` directly to the `to_msgpack` function as argument. Maybe the app already knows which type code to use.

In the original code the type code was saved as `_type` attribute. In my opinion the knowledge about the name of this special attribute should not be part of pyserde.
But with the `ext_type_code` as argument one could use it like this:
```python
d = DerivedA(i=1, s="a", j=10)
to_msgpack(d, ext_type_code=d._type)
```
So basically the app has to implement the lookup itself.

**Comment by adsharma** [comment]
The code that uses these APIs is here:

https://github.com/adsharma/raft/tree/master/raft/messages

I like the reversed dicts to keep the symmetry. The only reason why _type
exists is for serialization/pyserde.
We can discuss a different name if that's inconvenient.

File: serde/de.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yukinarit) [reply serde/de.py:399]: 👍

**Comment by yukinarit** [comment]
LGTM 👍

**Comment by yukinarit** [comment]
I will merge once [this thread](https://github.com/yukinarit/pyserde/pull/80#issuecomment-772836338) is resolved 👍

File: serde/msgpack.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ydylla) [reply serde/msgpack.py]: @yukinarit Thanks for the review.

I changed `to_msgpack` to use a `Dict[Type, int]` but I am not really happy with it
I guess it will confuse at least some users and also forces the application to save two dicts with basically the same information. But for know it is good enough, maybe someone else has a better idea in future.

Also keep in mind this is now a breaking change.

**Comment by yukinarit** [comment]
@adsharma Is everything ok? Could you check if your reviews were all addressed?

**Comment by adsharma** [comment]
Everything looks good and tests pass. Thanks for making the changes!
I need to push a small commit for the reversed dict once this is merged.

**Comment by ydylla** [comment]
Nice. Thanks @yukinarit and @adsharma.
